DRA: remove "source" indirection from v1 Pod API

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup
/kind api-change

#### What this PR does / why we need it:

This makes the API nicer:

    resourceClaims:
    - name: with-template
      resourceClaimTemplateName: test-inline-claim-template
    - name: with-claim
      resourceClaimName: test-shared-claim

Previously, this was:

    resourceClaims:
    - name: with-template
      source:
        resourceClaimTemplateName: test-inline-claim-template
    - name: with-claim
      source:
        resourceClaimName: test-shared-claim

A more long-term benefit is that other, future alternatives might not make sense under the "source" umbrella.

#### Special notes for your reviewer:

This is a breaking change. It's justified because DRA is still alpha and will have several other API breaks in 1.31.

#### Does this PR introduce a user-facing change?

```release-note
DRA: in the `pod.spec.recourceClaims` array, the `source` indirection is no longer necessary. Instead of e.g. `source: resourceClaimTemplateName: my-template`, one can write `resourceClaimTemplateName: my-template`.
```

pohly committed:
DRA: remove "source" indirection from v1 Pod API

This makes the API nicer:

    resourceClaims:
    - name: with-template
      resourceClaimTemplateName: test-inline-claim-template
    - name: with-claim
      resourceClaimName: test-shared-claim

Previously, this was:

    resourceClaims:
    - name: with-template
      source:
        resourceClaimTemplateName: test-inline-claim-template
    - name: with-claim
      source:
        resourceClaimName: test-shared-claim

A more long-term benefit is that other, future alternatives
might not make sense under the "source" umbrella.

This is a breaking change. It's justified because DRA is still
alpha and will have several other API breaks in 1.31.

cici37 commented:
/remove-sig api-machinery

fedebongio commented:
/remove-sig api-machinery

pohly commented:
/retest

seans3 commented:
/remove-sig api-machinery

pohly commented:
/retest

thockin reviewed (commented):
LGTM from API perspective

thockin commented:
Do you also want me to do full code review or @johnbelamaric can you help here?

johnbelamaric commented:
I can look in a bit but I am not an approver.

thockin commented:
If you are happy with the logic, I am happy with API, and I can do final
approve on most of it

On Fri, Jun 28, 2024 at 9:33 AM John Belamaric ***@***.***>
wrote:

> I can look in a bit but I am not an approver.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/125116#issuecomment-2197280408>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVH2FETRKX47FIIMTRTZJWF4VAVCNFSM6AAAAABI6DSUT2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCOJXGI4DANBQHA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

johnbelamaric reviewed (approved):
LGTM, changes are mostly mechanical, no logic changes.

/approve
/lgtm

pohly commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3020,35 +3020,28 @@ func validatePodResourceClaim(podMeta *metav1.ObjectMeta, claim core.PodResource
 			allErrs = append(allErrs, nameErrs...)
-		} else if podMeta != nil && claim.Source.ResourceClaimTemplateName != nil {
+		} else if podMeta != nil && claim.ResourceClaimTemplateName != nil {
 			claimName := podMeta.Name + "-" + claim.Name
```
We already stopped doing this deterministic naming thing several releases ago. I just forgot to remove this particular piece of validation for it :sweat: 

I'll fix this in this PR as I am touching these lines, even though it is a bit unrelated.

pohly replied:
Done, see additional commit.

I removed the validation test that was specifically for the combination of pod name + claim name. There are other test cases for the rest of the validation (uniqueness, bad name, etc.).

pohly committed:
api: remove obsolete validation of deterministic claim name

When replacing the deterministic ResourceClaim name with a generated one this
particular piece of the original validation was incorrectly left in place.
It's not required anymore that "<pod name>-<claim name in pod spec>" is a valid
ResourceClaim name.

thockin commented:
Thanks!

/lgtm
/approve

cici37 commented:
/triage accepted
